samples/nrf9160/modem_shell/src/link/link_shell.c

```diff
@@ -1682,6 +1682,387 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
+			break;
+		case 'r':
+			common_option = LINK_COMMON_READ;
+			break;
+		case LINK_SHELL_OPT_RESET:
+			common_option = LINK_COMMON_RESET;
+			break;
+		case LINK_SHELL_OPT_START:
+			common_option = LINK_COMMON_START;
+			break;
+		case LINK_SHELL_OPT_STOP:
+			common_option = LINK_COMMON_STOP;
+			break;
+		case '?':
+			/* TODO: We could print usage here */
+			break;
+		default:
+			break;
+		}
+	}
+	return common_option;
+}
+
+int link_funmode(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 0;
+	bool nmode_use_rel14 = true;
+	enum lte_lc_func_mode funmode_option = LINK_FUNMODE_NONE;
+	enum lte_lc_func_mode functional_mode;
+	enum link_shell_common_options common_option;
+	char snum[64];
+
+	common_option = link_shell_getopt_common(argc, argv);
+
+	/* Reset getopt due to possible previous failures */
+	optreset = 1;
+	/* We start from subcmd arguments */
+	optind = 1;
+
+	int long_index = 0;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv,
+				  "014",
+				  long_options, &long_index)) != -1) {
+		switch (opt) {
+		case '0':
+			funmode_option = LTE_LC_FUNC_MODE_POWER_OFF;
+			break;
+		case '1':
+			funmode_option = LTE_LC_FUNC_MODE_NORMAL;
+			break;
+		case '4':
+			funmode_option = LTE_LC_FUNC_MODE_OFFLINE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_LTEOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_LTE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_LTEON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_LTE;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_GNSSOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_GNSS;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_GNSSON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_GNSS;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_UICCOFF:
+			funmode_option = LTE_LC_FUNC_MODE_DEACTIVATE_UICC;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_UICCON:
+			funmode_option = LTE_LC_FUNC_MODE_ACTIVATE_UICC;
+			break;
+		case LINK_SHELL_OPT_FUNMODE_FLIGHTMODE_UICCON:
+			funmode_option = LTE_LC_FUNC_MODE_OFFLINE_UICC_ON;
+			break;
+		case LINK_SHELL_OPT_NMODE_NO_REL14:
+			funmode_option = LTE_LC_FUNC_MODE_NORMAL;
+			nmode_use_rel14 = false;
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (common_option == LINK_COMMON_READ) {
+		ret = lte_lc_func_mode_get(&functional_mode);
+		if (ret) {
+			mosh_error("Cannot get functional mode: %d", ret);
+		} else {
+			mosh_print(
+				"Functional mode read successfully: %s",
+				link_shell_funmode_to_string(functional_mode, snum));
+		}
+	} else if (funmode_option != LINK_FUNMODE_NONE) {
+		ret = link_func_mode_set(funmode_option, nmode_use_rel14);
+		if (ret < 0) {
+			mosh_error("Cannot set functional mode: %d", ret);
+		} else {
+			mosh_print(
+				"Functional mode set successfully: %s",
+				link_shell_funmode_to_string(
+					funmode_option, snum));
+		}
+	} else {
+		goto show_usage;
+	}
+	return ret;
+
+show_usage:
+	/* Reset getopt for another users: */
+	optreset = 1;
+
+	link_shell_print_usage(LINK_CMD_FUNMODE);
```

Comment:
Need to discuss also whether this is the way to print usage.
And I need to resolve where to call this.

Reply:
I guess we'll keep this style and `link_shell_print_usage` function